sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -404,9 +408,96 @@ private JmsIO() {}
     T mapMessage(Message message) throws Exception;
   }
 
+  /**
+   * Checkpoint for an unbounded JMS source. Consists of the JMS messages waiting to be acknowledged
+   * and oldest pending message timestamp.
+   */
+  @VisibleForTesting
+  static class JmsCheckpointMark implements UnboundedSource.CheckpointMark, Serializable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JmsCheckpointMark.class);
+
+    @VisibleForTesting Instant oldestMessageTimestamp = Instant.now();
+    @VisibleForTesting transient List<Message> messages = new ArrayList<>();
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    public JmsCheckpointMark() {}
+
+    public void add(Message message) throws Exception {
+      lock.writeLock().lock();
+      try {
+        Instant currentMessageTimestamp = new Instant(message.getJMSTimestamp());
+        if (currentMessageTimestamp.isBefore(oldestMessageTimestamp)) {
+          oldestMessageTimestamp = currentMessageTimestamp;
+        }
+        messages.add(message);
+      } finally {
+        lock.writeLock().unlock();
+      }
+    }
+
+    public Instant getOldestMessageTimestamp() {
+      lock.readLock().lock();
+      try {
+        return this.oldestMessageTimestamp;
+      } finally {
+        lock.readLock().unlock();
+      }
+    }
+
+    /**
+     * Acknowledge all outstanding message. Since we believe that messages will be delivered in
+     * timestamp order, and acknowledged messages will not be retried, the newest message in this
+     * batch is a good bound for future messages.
+     */
+    @Override
+    public void finalizeCheckpoint() {
+      lock.writeLock().lock();
+      try {
+        for (Message message : messages) {
+          try {
+            message.acknowledge();
+            Instant currentMessageTimestamp = new Instant(message.getJMSTimestamp());
+            if (currentMessageTimestamp.isAfter(oldestMessageTimestamp)) {
+              oldestMessageTimestamp = currentMessageTimestamp;
+            }
+          } catch (Exception e) {
+            LOG.error("Exception while finalizing message: {}", e);
+          }
+        }
+        messages.clear();
+      } finally {
+        lock.writeLock().unlock();
+      }
+    }
+
+    // set an empty list to messages when deserialize
+    private void readObject(java.io.ObjectInputStream stream)
+        throws IOException, ClassNotFoundException {
+      stream.defaultReadObject();
+      messages = new ArrayList<>();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof JmsCheckpointMark) {
+        JmsCheckpointMark that = (JmsCheckpointMark) other;
+        return Objects.equals(this.oldestMessageTimestamp, that.oldestMessageTimestamp);
```

Comment:
We need to add messages to the equals too.

Reply:
Messages list is transient. Do we really need it in the equals ?